### ISSUE TITLE
v3.3.0 (and v3.3.1) breaks test: Could not load file or assembly in method with AssemblyInitialize attribute

### ISSUE THREAD

**Issue by Andreas-Dorfer** [microsoft/testfx#2653]
I have the following test code that works with no issues up to v3.2.2, but fails with v3.3.0 and v3.3.1 (see [MsTestRegression.zip](https://github.com/microsoft/testfx/files/14872964/MsTestRegression.zip) for running example):

```C#
using FsCheck;

namespace MsTestRegression;

[TestClass]
public class TestClass
{
    [TestMethod]
    public void TestMethod1()
    {
        Prop.ForAll((IEnumerable<int> x) => true).QuickCheckThrowOnFailure();
    }

    [AssemblyInitialize]
    public static void TestInitialize(TestContext _)
    {
        Arb.Register<Arbitraries>();
    }
}

public class Arbitraries
{
    public static Arbitrary<IEnumerable<T>> EnumArb<T>() => new EnumerableArb<T>();
}

public sealed class EnumerableArb<T> : Arbitrary<IEnumerable<T>>
{
    public override Gen<IEnumerable<T>> Generator => Arb.Generate<List<T>>().Select(l => (IEnumerable<T>)l);
}
```
(I'm using [FsCheck](https://github.com/fscheck/FsCheck)).

When I debug the test with v3.2.2, there are no exceptions. However, when I debug the test with v3.3.0 (or v3.3.1), there are two exceptions thrown.

First:
![image](https://github.com/microsoft/testfx/assets/41114919/edaffaf5-283c-4281-866a-2ce7a1a8ce84)

Second:
![image](https://github.com/microsoft/testfx/assets/41114919/f363bf89-3cc4-4c0e-a0f1-d449ae5b7e3b)
![image](https://github.com/microsoft/testfx/assets/41114919/33c645f2-fbec-4be7-ac5f-cf8577a63087)

I think these exceptions cause the test to fail. It seems like something about loading assemblies in a method with the `AssemblyInitialize` attribute has changed with v3.3.0. The actual test method `TestMethod1` then fails because `Arb.Register<Arbitraries>()` doesn't work as expected.

The test also fails with v3.3.0 when run from the CLI (but works with up to v3.2.2).

*) .net8
*) MSTest.TestFramework 3.3.1
*) Visual Studio 17.9.5

**Comment by MarcoRossignoli** [comment]
I'm getting different exception in VS and also in console 
<img width="1217" alt="image" src="https://github.com/microsoft/testfx/assets/7894084/b4276e7e-2cd0-4bf7-81d3-94c9b18d47e3">

can you cleanup the solution removing all obj/bin folders?

**Comment by Andreas-Dorfer** [comment]
I do get that exception too with v3.3.0. But that's a subsequent error, because the registration of the arbitrary in `TestInitialize` doesn't work. That's what the exception message is saying: It cannot find an arbitrary for `IEnumerable<int>`, although I'm registrating the arbitrary for `IEnumerable<int>` in `TestInitialize`. When you downgrade to v3.2.2, the arbitrary registration works again, and the test runs successfully.

To reproduce the other 2 exceptions, you have to enable all "Common Language Runtime Exceptions" and uncheck "Enable Just My Code".

**Comment by Andreas-Dorfer** [comment]
I don't know if it helps you with evaluating the issue, but I'm maintaining a project that integrates [FsCheck with MsTest](https://github.com/Andreas-Dorfer/fscheck-mstest) and I use MsTest + FsCheck in production. Upgrading MsTest to v3.3.0 breaks a lot of my production tests that depend on arbitrary registration (both locally and in the CI pipeline).

**Comment by Evangelink** [comment]
I do confirm that with the repro and MSTest 3.2.2, the test pass while with 3.3.0 or 3.3.1 the test fails.

I have done some changes in local to try to revert #2110 and despite the few debugging exceptions all gone, I still end up with the same FsCheck exception: `System.Exception: The type System.Collections.Generic.IEnumerable`1[[System.Int32, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]] is not handled automatically by FsCheck. Consider using another type or writing and registering a generator for it.`

With or without the changes, we are correctly able to call the content of the `TestMethod1` so it would seem that the issue is on FsCheck. It's possible that some change in what is loaded causes FsCheck to fail but I don't know the logic they use.

I'd recommend creating a bug on their side so they can provide some more insights.

**Comment by osexpert** [comment]
I got the same\similar problem in 3.3.0-3.3.1.
It was a [TestMethod] with multiple [DataRow]. If I ran each [DataRow] individually, they worked. If I ran the [TestMethod], the first [DataRow] worked but the second one failed (weird COMException). So it's kind of a mystery error.
I downgraded to 2.2.10.

**Comment by Evangelink** [comment]
@osexpert are you also using FsCheck?

**Comment by osexpert** [comment]
> @osexpert are you also using FsCheck?

Not using FsCheck. But I tried to reproduce the problem again, I may not have remembered it clearly.....
When I tried to reproduce it again, it seems to work fine when tests succeed. But if a test fail and I try to debug it, then this exception is the first one I hit:

![image](https://github.com/microsoft/testfx/assets/33638596/63a3a605-653b-42dc-bea8-cc257105a231)
The exception will be catched thou, as the screenshot shows. But the comment is not right:
`// SetupConfiguration won't work if VS is not installed or VS is pre-vs2017 version.`
I have VS 2022.

So I think...this exception may not be harmful since catched, but it is confusing when debugging failing tests, so confusing\scary, that I downgraded instantly:-)

**Comment by Andreas-Dorfer** [comment]
@Evangelink I guess I found the issue. FsCheck's arbitrary registration uses `ThreadLocal`:

![image](https://github.com/microsoft/testfx/assets/41114919/30e219a0-d87a-4cb9-a45e-e943cff29398)

Up to v3.2.2, that is no issue as long as I don't run my tests in parallel. The `AssemblyInitialize` code and all my tests run on the same thread. However, starting with v3.3.0, the `AssemblyInitialize` code runs on a different thread than my tests - even when I don't parallelize my test run.

```CSharp
[assembly: DoNotParallelize]

namespace TestProject;

[TestClass]
public class UnitTest
{
    [AssemblyInitialize]
    public static void InitAssembly(TestContext _)
    {
        Console.WriteLine($"'AssemblyInitialize' thread id: {Environment.CurrentManagedThreadId}");
    }

    [ClassInitialize]
    public static void InitClass(TestContext _)
    {
        Console.WriteLine($"'ClassInitialize' thread id: {Environment.CurrentManagedThreadId}");
    }

    [TestMethod]
    public void TestMethod()
    {
        Console.WriteLine($"'TestMethod' thread id: {Environment.CurrentManagedThreadId}");
    }
}
```
Produces an output like this:
```
'AssemblyInitialize' thread id: 10
'ClassInitialize' thread id: 10
'TestMethod' thread id: 5
```

**Comment by Evangelink** [comment]
We need to use custom thread for that but this is IMO a new feature request. 

I understand this break is problematic but I am not super keen on considering this a bug/regression on our side as we are changing internals and the behavior wasn't a documented/supported behavior.

**Comment by nohwnd** [comment]
There is some good news. Good luck with your release!

**Comment by charlesdwmorrison** [comment]
I am seeing this issue in .Net 8.
Downgrading to MSTest3.2.2 fixed the problem for me as well.
Not sure why this is closed? What is the real fix? Stay on MSTest 3.2.2 forever?

**Comment by Evangelink** [comment]
Hi @charlesdwmorrison! 

Can you confirm you are also using FsCheck? If not, could you provide a repro or give us some details about your scenario?

FsCheck is relying on some internals of MSTest (not a documented or supported behavior) and we had to change this behavior to enable some new timeout features so we are stuck here.

@Andreas-Dorfer do you know when the new version of FsCheck will be released?

**Comment by osexpert** [comment]
I also think this case should not be closed. It throws there (VSInstallPath property) even if VS >= 2017 is installed. It did not before and it is in concflict with the comment in the code. So there is a regression there IMO. It "works", but this makes it annoying to debug tests, you always get into this ugly COMException first and it is not very trustworthy (bad first impression).

**Comment by Evangelink** [comment]
> It throws there (VSInstallPath property) even if VS >= 2017 is installed. It did not before and it is in concflict with the comment in the code. So there is a regression there IMO.

It cannot be considered as regression as it's some internal exception being caught. If you debug runtime or most of the tools you will see such behavior happening. Yet I agree this is not the best experience and that's why I fixed it in #2654 that will be available as part of the 3.4 release.

**Comment by MarcoRossignoli** [comment]
> I understand this break is problematic but I am not super keen on considering this a bug/regression on our side as we are changing internals and the behavior wasn't a documented/supported behavior.

I would not consider a breaking change, MSTest doesn't give any guarantee for what I know(I don't see documentation around it) on the nature of the threads that are running the "tests method", we only guarantee that the execution order is the expected one. 
Tools/Tests should not take "dependency" on the runtime non documented/guarantee context.
